### DISCUSSION TITLE
Using Kobweb with NTLM proxy

### DISCUSSION THREAD

**Discussion by iainism** [varabyte/kobweb#617] [Q&A] [answered]
Hi all,

Does anyone know the 'secret sauce' that will allow Kobweb to build and run when behind a corporate NTLM proxy please?

So far I have added my corporate proxy's signing certificate to cacerts (as our proxy performs a certificate insertion), added the proxy settings in the Idea settings ('Connection was successful'), and added proxy information within the project-level `gradle.properties`, e.g.

```
systemProp.http.proxyHost=proxy.path.net
systemProp.http.proxyPort=123
systemProp.https.proxyHost=proxy.path.net
systemProp.https.proxyPort=123
```

However when trying to build a new, empty project, I get the result:
```
> Task :kotlinNpmInstall FAILED
(node:29168) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
error Couldn't find package "format-util@^1.0.5" required by "youtube@1.0.0-SNAPSHOT" on the "npm" registry.
Error: Couldn't find package "format-util@^1.0.5" required by "youtube-test@1.0.0-SNAPSHOT" on the "npm" registry.
    at MessageError.ExtendableBuiltin (C:\Users\myUser\.gradle\yarn\yarn-v1.22.17\lib\cli.js:721:66)
    at new MessageError (C:\Users\myUser\.gradle\yarn\yarn-v1.22.17\lib\cli.js:750:123)
    at PackageRequest.<anonymous> (C:\Users\myUser\.gradle\yarn\yarn-v1.22.17\lib\cli.js:36582:17)
    at Generator.throw (<anonymous>)
Error: Couldn't find package "format-util@^1.0.5" required by "youtube-test@1.0.0-SNAPSHOT" on the "npm" registry.

    at step (C:\Users\myUser\.gradle\yarn\yarn-v1.22.17\lib\cli.js:310:30)
    at C:\Users\myUser\.gradle\yarn\yarn-v1.22.17\lib\cli.js:323:13
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

[... snippety snip ...]

```
Error: Couldn't find package "karma-chrome-launcher@3.2.0" required by "youtube-test@1.0.0-SNAPSHOT" on the "npm" registry.

Error: Couldn't find package "mocha@10.7.0" required by "youtube-test@1.0.0-SNAPSHOT" on the "npm" registry.

Error: Couldn't find package "karma-webpack@5.0.1" required by "youtube-test@1.0.0-SNAPSHOT" on the "npm" registry.


Execution failed for task ':kotlinNpmInstall'.
>                 Process 'Resolving NPM dependencies using yarn' returns 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
BUILD FAILED in 12s
21 actionable tasks: 6 executed, 15 up-to-date
Configuration cache entry stored.
```

Is there something else I should be doing please?

**Comment by iainism** [comment, marked as answer]
Ah-ha! I needed to tell NPM that it should use an extra certificate for chain of trust, as described here: https://stackoverflow.com/a/56239954/955198

I permanently added a (Windows) system variable, NODE_EXTRA_CA_CERTS, with the value set to a downloaded version of our proxy's root signing certificate in PEM format.

The build now runs to completion.

**Reply by iainism** [reply]
Note that setting the proxy properties (e.g. `systemProp.http.proxyHost=proxy.path.net`) in the `gradle.properties` is *not* required!

**Reply by bitspittle** [reply]
Sorry I couldn't share an answer here, I had no idea! Thank you for pushing through and sharing your update with us here.

**Reply by iainism** [reply]
My pleasure. Is it worth trying to summarise this somewhere in the documentation to save time for others that might hit the same issue?

**Reply by bitspittle** [reply]
I'm not sure. You're the first person I've heard bring it up. And now this discussion exists which I hope people can find? 

Of course if any user comes to this discussion because they had the same issue, please let me know!
